File: blue-core/src/blue/blueLive/LiveObjectSet.java
/*
 * blue - object composition environment for csound
 * Copyright (c) 2012 Steven Yi (stevenyi@gmail.com)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by  the Free Software Foundation; either version 2 of the License or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.LIB.  If not, write to
 * the Free Software Foundation Inc., 59 Temple Place - Suite 330,
 * Boston, MA  02111-1307 USA
 */

/**
 * Class to hold sets of LiveObjects (really is an instance of ArrayList, but 
 * called Set here as that is what it is called by users)
 */
package blue.blueLive;

import electric.xml.Element;
import electric.xml.Elements;
import java.util.ArrayList;

/**
 *
 * @author stevenyi
 */
public class LiveObjectSet extends ArrayList<LiveObject> {
    
    public String name = "";

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }
    
    public Element saveAsXML() {
         Element retVal = new Element("liveObjectSet");
         retVal.setAttribute("name", name);
         
         for(LiveObject liveObj : this) {
             retVal.addElement("liveObjectRef").setText(liveObj.getUniqueId());
         }

         return retVal;
    }
    
    public static LiveObjectSet loadFromXML(Element data, LiveObjectBins bins) {
        LiveObjectSet retVal = new LiveObjectSet();
        
        String val = data.getAttributeValue("name");
        if (val != null && val.length() > 0) {
            retVal.name = val;
        }
        
        Elements nodes = data.getElements();
        
        while (nodes.hasMoreElements()) {

            Element node = nodes.next();
            String nodeName = node.getName();

            if (nodeName.equals("liveObjectRef")) {
                String uniqueId = node.getTextString();
                LiveObject lObj = bins.getLiveObjectByUniqueId(uniqueId);
                
                if(lObj != null) {
                    retVal.add(lObj);
                }
            }
            
        }
        
        return retVal;
    }
    
}

File: packages/blue-data/src/live/live-object-set.ts
import { Element } from '../serialization/xml-reader';
import { BlueDataObject } from '../blue-data-object';
import { LiveObject } from './live-object';
import { LiveObjectBins } from './live-object-bins';

export class LiveObjectSet implements BlueDataObject {
  private _name = '';
  private _liveObjectIds: string[] = [];

  constructor() {}

  getName(): string {
    return this._name;
  }

  setName(name: string): void {
    this._name = name;
  }

  getLiveObjectIds(): string[] {
    return this._liveObjectIds;
  }

  setLiveObjectIds(ids: string[]): void {
    this._liveObjectIds = ids;
  }

  saveAsXML(): Element {
    const elem = new Element('liveObjectSet');
    elem.setAttribute('name', this._name);
    for (const id of this._liveObjectIds) {
      elem.addElement('liveObjectRef').setText(id);
    }
    return elem;
  }

  static loadFromXML(data: Element, bins: LiveObjectBins): LiveObjectSet {
    const set = new LiveObjectSet();
    const val = data.getAttribute('name');
    if (val) {
      set._name = val;
    }

    // Retain all saved-set identifiers losslessly, including references to
    // LiveObjects that no longer exist. Java Blue discarded missing IDs on
    // load; this parity pass retains them so legacy projects round-trip
    // without silent data loss. Resolution against existing objects happens
    // only when the set is applied (see resolveLiveObjects).
    const ids: string[] = [];
    const nodes = data.getElements();
    while (nodes.hasMoreElements()) {
      const node = nodes.next();
      if (node.getName() === 'liveObjectRef') {
        const uniqueId = node.getTextString();
        if (uniqueId) {
          ids.push(uniqueId);
        }
      }
    }
    set._liveObjectIds = ids;
    return set;
  }

  deepCopy(): BlueDataObject {
    const copy = new LiveObjectSet();
    copy._name = this._name;
    copy._liveObjectIds = [...this._liveObjectIds];
    return copy;
  }

  resolveLiveObjects(bins: LiveObjectBins): LiveObject[] {
    const result: LiveObject[] = [];
    for (const id of this._liveObjectIds) {
      const obj = bins.getLiveObjectByUniqueId(id);
      if (obj) result.push(obj);
    }
    return result;
  }
}
